src/data/finset/noncomm_prod.lean

```diff
@@ -183,6 +183,14 @@ given a proof that `+` commutes on all elements `f x` for `x ∈ s`."]
 def noncomm_prod (s : finset α) (f : α → β) (comm : ∀ (x ∈ s) (y ∈ s), commute (f x) (f y)) : β :=
 (s.1.map f).noncomm_prod (by simpa [multiset.mem_map, ←finset.mem_def] using comm)
 
+/-- A helper for rewriting the dependent `finset` argument together with the commutativity proof -/
+@[to_additive "A helper for rewriting the dependnet `finset` argument together with the
+commutativity proof"]
+lemma noncomm_prod_congr
+  {s₁ s₂ : finset α} (h : s₁ = s₂) (f : α → β)
+  (comm : ∀ (x ∈ s₁) (y ∈ s₁), commute (f x) (f y)) :
+  noncomm_prod s₁ f comm = noncomm_prod s₂ f (h ▸ comm) := by subst h
```

Comment:
It might be sensible to have this take distinct f and g, and require a proof they are equal on the set

Reply:
Hmm, would this still allow a simple
```
   rw finset.noncomm_prod_congr h,
```
then? Maybe a separate lemma for changing `f` is useful?

Reply:
I'm thinking of finset.prod_congr, by comparison

Reply:
Right, but that wouldn’t be (as) convenient to use when rewriting.

I guess the situation is similar as to `congr_arg` and `congr_fun`.

Should I rename this one to `noncomm_prod_congr_finset`, and add the most general `noncomm_prod_congr` that allows changing both?

Reply:
I did that, and added the obvious `multiset.noncomm_prod_congr` lemma while we are at it.

Reply:
I'm not really convinced `noncomm_prod_congr_finset` is useful you can just use `simp_rw h`

Reply:
But I can't: If I change the line
```
   rw finset.noncomm_prod_congr h,
```
in https://github.com/leanprover-community/mathlib/pull/11744/files#diff-dbff33c4f941f049b533b60208287dccc7a966ba9123603f9e09b8ea5ab6dbdeR98 to
```
  simp_rw h,
```
I get
```
invalid simplification lemma 'h' (use command 'set_option trace.simp_lemmas true' for more details)
```
and
```
   rw h
```
(which I intuitively would like to do) yields
```
rewrite tactic failed, motive is not type correct
  λ (_a : finset ι),
    finset.univ.noncomm_prod (λ (j : ι), ⇑(ϕ j) (pi.mul_single i y j)) _ = ⇑(ϕ i) y =
      (_a.noncomm_prod (λ (j : ι), ⇑(ϕ j) (pi.mul_single i y j)) _ = ⇑(ϕ i) y)
```

Reply:
Thanks for the example. That one is tricky, because `h` has `univ` on both sides, so `simp_rw` says "I can't use that, it would loop!". Maybe you're right and we do need these lemmas - although it would be nice to have a thread on zulip about how to deal with dependent rewrites which aren't appropriate for `simp_rw`.

Reply:
This works in your other PR:
```
  have h : insert i (finset.univ.erase i) = finset.univ:=
     (finset.insert_erase (finset.mem_univ i)),
  simp only [←h] {single_pass := tt},
```

Reply:
Ah, the problem is that the `h`, while good for simplification, is not suitable for simplification, because it’s LHS appears on the RHS, and then `simp_rw` fails.

Maybe I can work around that… but I really wish that `rw` would just be smarter and things would just work™…

Reply:
Indeed, that was it. Ok, this obliterates the need for this PR for my work. I’ll trim it down to the (supposedly) useful congr lemma.

Reply:
Ah, you also found that. Thanks for the even simpler work-around than mine!